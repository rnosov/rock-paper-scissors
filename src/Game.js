/*
 * Game React Component
 *
 * Copyright Â© Roman Nosov 2016
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */
import React, { Component, PropTypes } from 'react';
import 'animate.css/source/_base.css';
import 'animate.css/source/attention_seekers/bounce.css';
import 'animate.css/source/fading_entrances/fadeIn.css';

const
  propTypes = {
		elements: PropTypes.array,
		compare: PropTypes.func,
		translateResult: PropTypes.func,
  },
  DRAW = -1, WIN = 0, LOSS = 1,
  defaultProps = {
		elements: ['Rock', 'Paper', 'Scissors'],
		compare: (elements, x, y) => {
			x = elements.indexOf(x);
			y = elements.indexOf(y);
			if ( x === y )
				return DRAW;			
			const c = (x - y) % elements.length, mod = c < 0 ? c + elements.length : c;
			return mod < elements.length / 2 ? WIN : LOSS;
		},
		translateResult: result => result === DRAW ? 'Draw' : result === LOSS ? 'You lost' : 'You won',	
  };

class Game extends Component {	
	
	constructor() {
		super();		
		this.state = { rounds: [], isAnimationPlaying: false };
		this.onClick = this.handleClick.bind(this);
		this.onReset = this.handleReset.bind(this);
		this.onSimulate = this.handleSimulate.bind(this);
	}

	simulateMove() {
		return this.props.elements[Math.floor(Math.random()*this.props.elements.length)];
	}

	play(...moves) {	
		if (this.state.isAnimationPlaying) 
			return;
		const result = this.props.compare(this.props.elements, ...moves );
		this.setState({ isAnimationPlaying: true });		
		setTimeout( () => this.setState({ isAnimationPlaying: false, rounds: [{ result, moves }, ...this.state.rounds] }), 1000 );
	}

	handleClick({ currentTarget }) {
		this.play(currentTarget.getAttribute('data-id'), this.simulateMove());
	}

	handleReset() {
		this.setState({ rounds: [] });
	}

	handleSimulate() {
		this.play(this.simulateMove(), this.simulateMove());
	}

	displayScoreboard() {
		return (
			<div>
				<div className="row">
					<button 
      			onClick={this.onReset} 
      			type="button" 
      			className={`btn btn-outline-danger${this.state.rounds.length?'':' disabled'}`}
      		>
      			Reset Game
      		</button>
				</div>
				<br/>
				<div className="row">
					<ul className="list-group text-xs-left">
						{this.state.isAnimationPlaying? <li className="list-group-item active">Playing</li> : void 0 }
						{this.state.rounds.map( (round, index) => 
							<li className={`list-group-item ${ !index && !this.state.isAnimationPlaying ? 'active':'hidden-xs-down'}`} key={index}>
								{this.props.translateResult(round.result)} ( {round.moves[0]} - {round.moves[1]} ) 
							</li> 
						)}
					</ul>
				</div>
			</div>
		);
	}

	displayMove(index) {		
  	const score = this.state.rounds.reduce( (score, { result }) => score + ( result === index ? 1 : 0 ), 0);
		return (
			<div className="row">			
				<div key={this.state.rounds.length} className={ `animated ${this.state.isAnimationPlaying ? 'bounce' : 'fadeIn' }` }>
					<i 
						style={{ fontSize: '9em' }} 
						className={ this.state.isAnimationPlaying || !this.state.rounds.length ? 'icon-rock' : `icon-${this.state.rounds[0].moves[index].toLowerCase()}` }
					>
					</i>				
				</div>
				<h3>Score: {score}</h3>
      </div>
		);
	}

	displayPlayerUI() {
		return(
			<div className="row">
			  {this.props.elements.map( el => 			   	
      		<button 
      			key={el}
      			data-id={el}
      			onClick={this.onClick} 
      			type="button" 
      			className={`btn btn-outline-primary${ this.state.isAnimationPlaying ? ' disabled' : '' }`}
      		>
      			<i className={`icon-${el.toLowerCase()}`}></i> {el}
      		</button>      			
			  )}
      </div>
    );
  }

	displaySimulationButton() {
		return(	
  		<div className="row">
      	<button 
      		onClick={this.onSimulate} 
      		type="button"
      		className={`btn btn-outline-warning${ this.state.isAnimationPlaying ? ' disabled' : '' }`}      					
      	>
      		Simulate My Move
      	</button>
      </div>
    );
  }

	render() {
    return(
    	<div className="container">
			  <div className="row">
			    <div className="col-sm-4">
			    	<h2>You</h2>
			    	{this.displayPlayerUI()}
      			{this.displayMove(0)}
			    </div>
			    <div className="col-sm-4">
    			  <h2>{this.state.rounds.length} round{this.state.rounds.length === 1 ? '':'s'} played</h2>
			      {this.displayScoreboard()}
			    </div>
			    <div className="col-sm-4">
			    	<h2>Computer</h2>
			    	{this.displaySimulationButton()}
      			{this.displayMove(1)}
			    </div>
			  </div>
			</div>
    );
  }
}

Game.propTypes = propTypes;
Game.defaultProps = defaultProps;
export default Game;