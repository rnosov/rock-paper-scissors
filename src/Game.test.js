import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import Game from './Game';

const DRAW = -1, WIN = 0, LOSS = 1;

it('renders a initial view', () => {
  const instance = mount(<Game />);
  expect(instance.html()).toMatchSnapshot();
});

it('starts animation', () => {
  const instance = mount(<Game />);
	instance.find('button').first().simulate('click');
	expect(instance.state().isAnimationPlaying).toBe(true);
 	expect(instance.html()).toMatchSnapshot();
});

it('renders results correctly', () => {
  const instance = mount(<Game />);
	instance.setState({rounds: [
		{ result: LOSS, moves: ['Rock', 'Paper'] }, 
		{ result: WIN, moves: ['Scissors', 'Paper'] },
		{ result: WIN, moves: ['Paper', 'Rock'] },
	]});
  expect(instance.html()).toMatchSnapshot();
});

it('compares correctly', () => {
	const { elements, compare } = Game.defaultProps;
	expect(compare(elements,'Rock','Paper')).toBe(LOSS);  
	expect(compare(elements,'Rock','Scissors')).toBe(WIN);  
	expect(compare(elements,'Rock','Rock')).toBe(DRAW);  
	
	expect(compare(elements,'Paper','Paper')).toBe(DRAW);  
	expect(compare(elements,'Paper','Scissors')).toBe(LOSS);  
	expect(compare(elements,'Paper','Rock')).toBe(WIN);  

	expect(compare(elements,'Scissors','Paper')).toBe(WIN);  
	expect(compare(elements,'Scissors','Scissors')).toBe(DRAW);  
	expect(compare(elements,'Scissors','Rock')).toBe(LOSS);  
});

it('translates correctly', () => {
	expect(Game.defaultProps.translateResult(DRAW)).toBe('Draw');  
	expect(Game.defaultProps.translateResult(WIN)).toBe('You won');  
	expect(Game.defaultProps.translateResult(LOSS)).toBe('You lost');  
});
