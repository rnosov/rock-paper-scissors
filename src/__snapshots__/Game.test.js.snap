exports[`test renders a initial view 1`] = `"<div class=\"container\"><div class=\"row\"><div class=\"col-sm-4\"><h2>You</h2><div class=\"row\"><button data-id=\"Rock\" type=\"button\" class=\"btn btn-outline-primary\"><i class=\"icon-rock\"></i><!-- react-text: 8 --> <!-- /react-text --><!-- react-text: 9 -->Rock<!-- /react-text --></button><button data-id=\"Paper\" type=\"button\" class=\"btn btn-outline-primary\"><i class=\"icon-paper\"></i><!-- react-text: 12 --> <!-- /react-text --><!-- react-text: 13 -->Paper<!-- /react-text --></button><button data-id=\"Scissors\" type=\"button\" class=\"btn btn-outline-primary\"><i class=\"icon-scissors\"></i><!-- react-text: 16 --> <!-- /react-text --><!-- react-text: 17 -->Scissors<!-- /react-text --></button></div><div class=\"row\"><div class=\"animated fadeIn\"><i class=\"icon-rock\" style=\"font-size: 9em;\"></i></div><h3><!-- react-text: 22 -->Score: <!-- /react-text --><!-- react-text: 23 -->0<!-- /react-text --></h3></div></div><div class=\"col-sm-4\"><h2><!-- react-text: 26 -->0<!-- /react-text --><!-- react-text: 27 --> round<!-- /react-text --><!-- react-text: 28 -->s<!-- /react-text --><!-- react-text: 29 --> played<!-- /react-text --></h2><div><div class=\"row\"><button type=\"button\" class=\"btn btn-outline-danger disabled\">Reset Game</button></div><br><div class=\"row\"><ul class=\"list-group text-xs-left\"></ul></div></div></div><div class=\"col-sm-4\"><h2>Computer</h2><div class=\"row\"><button type=\"button\" class=\"btn btn-outline-warning\">Simulate My Move</button></div><div class=\"row\"><div class=\"animated fadeIn\"><i class=\"icon-rock\" style=\"font-size: 9em;\"></i></div><h3><!-- react-text: 44 -->Score: <!-- /react-text --><!-- react-text: 45 -->0<!-- /react-text --></h3></div></div></div></div>"`;

exports[`test renders results correctly 1`] = `"<div class=\"container\"><div class=\"row\"><div class=\"col-sm-4\"><h2>You</h2><div class=\"row\"><button data-id=\"Rock\" type=\"button\" class=\"btn btn-outline-primary\"><i class=\"icon-rock\"></i><!-- react-text: 8 --> <!-- /react-text --><!-- react-text: 9 -->Rock<!-- /react-text --></button><button data-id=\"Paper\" type=\"button\" class=\"btn btn-outline-primary\"><i class=\"icon-paper\"></i><!-- react-text: 12 --> <!-- /react-text --><!-- react-text: 13 -->Paper<!-- /react-text --></button><button data-id=\"Scissors\" type=\"button\" class=\"btn btn-outline-primary\"><i class=\"icon-scissors\"></i><!-- react-text: 16 --> <!-- /react-text --><!-- react-text: 17 -->Scissors<!-- /react-text --></button></div><div class=\"row\"><div class=\"animated fadeIn\"><i class=\"icon-rock\" style=\"font-size: 9em;\"></i></div><h3><!-- react-text: 22 -->Score: <!-- /react-text --><!-- react-text: 23 -->2<!-- /react-text --></h3></div></div><div class=\"col-sm-4\"><h2><!-- react-text: 26 -->3<!-- /react-text --><!-- react-text: 27 --> round<!-- /react-text --><!-- react-text: 28 -->s<!-- /react-text --><!-- react-text: 29 --> played<!-- /react-text --></h2><div><div class=\"row\"><button type=\"button\" class=\"btn btn-outline-danger\">Reset Game</button></div><br><div class=\"row\"><ul class=\"list-group text-xs-left\"><li class=\"list-group-item active\"><!-- react-text: 49 -->You lost<!-- /react-text --><!-- react-text: 50 --> ( <!-- /react-text --><!-- react-text: 51 -->Rock<!-- /react-text --><!-- react-text: 52 --> - <!-- /react-text --><!-- react-text: 53 -->Paper<!-- /react-text --><!-- react-text: 54 --> )<!-- /react-text --></li><li class=\"list-group-item hidden-xs-down\"><!-- react-text: 56 -->You won<!-- /react-text --><!-- react-text: 57 --> ( <!-- /react-text --><!-- react-text: 58 -->Scissors<!-- /react-text --><!-- react-text: 59 --> - <!-- /react-text --><!-- react-text: 60 -->Paper<!-- /react-text --><!-- react-text: 61 --> )<!-- /react-text --></li><li class=\"list-group-item hidden-xs-down\"><!-- react-text: 63 -->You won<!-- /react-text --><!-- react-text: 64 --> ( <!-- /react-text --><!-- react-text: 65 -->Paper<!-- /react-text --><!-- react-text: 66 --> - <!-- /react-text --><!-- react-text: 67 -->Rock<!-- /react-text --><!-- react-text: 68 --> )<!-- /react-text --></li></ul></div></div></div><div class=\"col-sm-4\"><h2>Computer</h2><div class=\"row\"><button type=\"button\" class=\"btn btn-outline-warning\">Simulate My Move</button></div><div class=\"row\"><div class=\"animated fadeIn\"><i class=\"icon-paper\" style=\"font-size: 9em;\"></i></div><h3><!-- react-text: 44 -->Score: <!-- /react-text --><!-- react-text: 45 -->1<!-- /react-text --></h3></div></div></div></div>"`;

exports[`test starts animation 1`] = `"<div class=\"container\"><div class=\"row\"><div class=\"col-sm-4\"><h2>You</h2><div class=\"row\"><button data-id=\"Rock\" type=\"button\" class=\"btn btn-outline-primary disabled\"><i class=\"icon-rock\"></i><!-- react-text: 8 --> <!-- /react-text --><!-- react-text: 9 -->Rock<!-- /react-text --></button><button data-id=\"Paper\" type=\"button\" class=\"btn btn-outline-primary disabled\"><i class=\"icon-paper\"></i><!-- react-text: 12 --> <!-- /react-text --><!-- react-text: 13 -->Paper<!-- /react-text --></button><button data-id=\"Scissors\" type=\"button\" class=\"btn btn-outline-primary disabled\"><i class=\"icon-scissors\"></i><!-- react-text: 16 --> <!-- /react-text --><!-- react-text: 17 -->Scissors<!-- /react-text --></button></div><div class=\"row\"><div class=\"animated bounce\"><i class=\"icon-rock\" style=\"font-size: 9em;\"></i></div><h3><!-- react-text: 22 -->Score: <!-- /react-text --><!-- react-text: 23 -->0<!-- /react-text --></h3></div></div><div class=\"col-sm-4\"><h2><!-- react-text: 26 -->0<!-- /react-text --><!-- react-text: 27 --> round<!-- /react-text --><!-- react-text: 28 -->s<!-- /react-text --><!-- react-text: 29 --> played<!-- /react-text --></h2><div><div class=\"row\"><button type=\"button\" class=\"btn btn-outline-danger disabled\">Reset Game</button></div><br><div class=\"row\"><ul class=\"list-group text-xs-left\"><li class=\"list-group-item active\">Playing</li></ul></div></div></div><div class=\"col-sm-4\"><h2>Computer</h2><div class=\"row\"><button type=\"button\" class=\"btn btn-outline-warning disabled\">Simulate My Move</button></div><div class=\"row\"><div class=\"animated bounce\"><i class=\"icon-rock\" style=\"font-size: 9em;\"></i></div><h3><!-- react-text: 44 -->Score: <!-- /react-text --><!-- react-text: 45 -->0<!-- /react-text --></h3></div></div></div></div>"`;
